dbms/src/Storages/KVStore/Utils/AsyncTasks.h

```diff
@@ -125,9 +310,20 @@ struct AsyncTasks
     }
 
 protected:
-    std::unordered_map<Key, std::future<R>> futures;
-    std::unordered_map<Key, uint64_t> start_time;
+    std::shared_ptr<CancelHandle> getCancelHandleFromCaller(Key k) const
+    {
+        std::unique_lock<std::mutex> l(mtx);
+        auto it = tasks.find(k);
+        RUNTIME_CHECK_MSG(it != tasks.end(), "getCancelHandleFromCaller meets empty key");
+        return it->second.cancel;
+    }
+
+
+protected:
+    std::unordered_map<Key, Elem> tasks;
+    // TODO(fap) Use threadpool which supports purging from queue.
     std::unique_ptr<ThreadPool> thread_pool;
     mutable std::mutex mtx;
+    LoggerPtr log;
```

Comment:
It is not used currently, however, leave it here, it could be useful when debugging some cases.